db/sql/00_msar.sql

```diff
@@ -3228,3 +3228,222 @@ BEGIN
+SELECT jsonb_agg(jsonb_build_object('attnum', attnum, 'direction', 'asc'))
+FROM pg_constraint, LATERAL unnest(conkey) attnum
+WHERE contype='p' AND conrelid=tab_id AND has_column_privilege(tab_id, attnum, 'SELECT');
+$$ LANGUAGE SQL RETURNS NULL ON NULL INPUT;
+
+
+CREATE OR REPLACE FUNCTION msar.get_total_order(tab_id oid) RETURNS jsonb AS $$
+WITH orderable_cte AS (
+  SELECT attnum
+  FROM pg_catalog.pg_attribute
+    INNER JOIN pg_catalog.pg_cast ON atttypid=castsource
+    INNER JOIN pg_catalog.pg_operator ON casttarget=oprleft
+  WHERE
+    attrelid=tab_id
+    AND attnum>0
+    AND castcontext='i'
+    AND oprname='<'
+  UNION SELECT attnum
+  FROM pg_catalog.pg_attribute
+    INNER JOIN pg_catalog.pg_operator ON atttypid=oprleft
+  WHERE
+    attrelid=tab_id
+    AND attnum>0
+    AND oprname='<'
+  ORDER BY attnum
+)
+SELECT COALESCE(jsonb_agg(jsonb_build_object('attnum', attnum, 'direction', 'asc')), '[]'::jsonb)
+-- This privilege check is redundant in context, but may be useful for other callers.
+FROM orderable_cte
+-- This privilege check is redundant in context, but may be useful for other callers.
+WHERE has_column_privilege(tab_id, attnum, 'SELECT');
+$$ LANGUAGE SQL RETURNS NULL ON NULL INPUT;
+
+
+CREATE OR REPLACE FUNCTION
+msar.build_order_by_expr(tab_id oid, order_ jsonb) RETURNS text AS $$/*
+Build an ORDER BY expression for the given table and order JSON.
+
+The ORDER BY expression will refer to columns by their attnum. This is designed to work together
+with `msar.build_selectable_column_expr`. It will only use the columns to which the user has access.
+Finally, this function will append either a primary key, or all columns to the produced ORDER BY so
+the resulting ordering is totally defined (i.e., deterministic).
+
+Args:
+  tab_id: The OID of the table whose columns we'll order by.
+*/
+SELECT 'ORDER BY ' || string_agg(format('%I %s', attnum, msar.sanitize_direction(direction)), ', ')
+FROM jsonb_to_recordset(
+    COALESCE(
+      COALESCE(order_, '[]'::jsonb) || msar.get_pkey_order(tab_id),
+      COALESCE(order_, '[]'::jsonb) || msar.get_total_order(tab_id)
+    )
+)
+  AS x(attnum smallint, direction text)
+WHERE has_column_privilege(tab_id, attnum, 'SELECT');
+$$ LANGUAGE SQL;
+
+
+CREATE OR REPLACE FUNCTION
+msar.build_selectable_column_expr(tab_id oid) RETURNS text AS $$/*
+Build an SQL select-target expression of only columns to which the user has access.
+
+Given columns with attnums 2, 3, and 4, and assuming the user has access only to columns 2 and 4,
+this function will return an expression of the form:
+
+column_name AS "2", another_column_name AS "4"
+
+Args:
+  tab_id: The OID of the table containing the columns to select.
+*/
+SELECT string_agg(format('msar.format_data(%I) AS %I', attname, attnum), ', ')
+FROM pg_catalog.pg_attribute
+WHERE
+  attrelid = tab_id AND attnum > 0 AND has_column_privilege(attrelid, attnum, 'SELECT');
```

Comment:
This breaks for a table from which previously existing column(s) have been dropped.
Here is the specific error:
```py
mathesar=# select msar.list_records_from_table('"Library Management"."Items"'::regclass::oid, NULL, NULL, NULL, NULL, NULL, NULL);
ERROR:  column "........pg.dropped.4........" does not exist
LINE 5: ...data("Acquisition Date") AS "3", msar.format_data("........p...
                                                             ^
QUERY:  
    WITH count_cte AS (
      SELECT count(1) AS count FROM "Library Management"."Items"
    ), results_cte AS (
      SELECT msar.format_data(id) AS "1", msar.format_data("Barcode") AS "2", msar.format_data("Acquisition Date") AS "3", msar.format_data("........pg.dropped.4........") AS "4", msar.format_data("Book") AS "5" FROM "Library Management"."Items" ORDER BY "1" ASC LIMIT NULL OFFSET NULL
    )
    SELECT jsonb_build_object(
      'results', jsonb_agg(row_to_json(results_cte.*)),
      'count', max(count_cte.count)
    )
    FROM results_cte, count_cte
    
CONTEXT:  PL/pgSQL function msar.list_records_from_table(oid,integer,integer,jsonb,jsonb,jsonb,jsonb) line 19 at EXECUTE
```

Reply:
Nice catch! I've pushed a fix and a test. Should be fine now.